File: services/downloader/schema.py
import uuid
import enum
from sqlalchemy import (
    Column, String, Integer, Float, Boolean, DateTime, ForeignKey, 
    Enum, Index, UniqueConstraint, BigInteger
)
from sqlalchemy.dialects.postgresql import UUID, ARRAY
from sqlalchemy.orm import declarative_base, relationship
from sqlalchemy.sql import func

Base = declarative_base()

# ---------------------------------------------------------
# ENUMS
# ---------------------------------------------------------
class DownloadState(enum.Enum):
    IDLE = "IDLE"
    FETCHING_METADATA = "FETCHING_METADATA"
    DOWNLOADING = "DOWNLOADING"
    PROCESSING = "PROCESSING"
    COMPLETED = "COMPLETED"
    ERROR = "ERROR"

class StorageProvider(enum.Enum):
    NONE = "NONE"
    LOCAL_DISK = "LOCAL_DISK"
    MINIO = "MINIO"

class SourceType(enum.Enum):
    MAGNET = "MAGNET"
    TORRENT_FILE = "TORRENT_FILE"

# ---------------------------------------------------------
# MOVIES TABLE
# ---------------------------------------------------------
class Movie(Base):
    __tablename__ = "movies"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    yts_id = Column(Integer, unique=True, nullable=False)
    
    title = Column(String, nullable=False)
    title_long = Column(String)
    release_year = Column(Integer, nullable=False)
    rating = Column(Float)
    runtime_minutes = Column(Integer)
    description = Column(String)
    genres = Column(ARRAY(String))
    language = Column(String)
    mpa_rating = Column(String)

    yts_poster_url = Column(String, nullable=False)
    minio_poster_path = Column(String)
    yts_background_image_url = Column(String)
    minio_background_image_path = Column(String)

    imdb_code = Column(String)
    date_uploaded = Column(DateTime(timezone=True))

    # --- NEW FIELDS ---
    is_new_release = Column(Boolean, default=False)
    is_popular = Column(Boolean, default=False)
    
    last_checked_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False)

    # Relationships
    torrents = relationship("Torrent", back_populates="movie", cascade="all, delete-orphan")
    downloads = relationship("Download", back_populates="movie", cascade="all, delete-orphan")
    playbacks = relationship("UserPlayback", back_populates="movie", cascade="all, delete-orphan")
    directors = relationship("MovieDirector", back_populates="movie", cascade="all, delete-orphan")
    cast = relationship("MovieCast", back_populates="movie", cascade="all, delete-orphan")
    watch_history_entries = relationship("WatchHistory", back_populates="movie", cascade="all, delete-orphan")
    favorites = relationship("UserFavorite", back_populates="movie", cascade="all, delete-orphan")
    comments = relationship("MovieComment", back_populates="movie", cascade="all, delete-orphan")

    __table_args__ = (
        Index("yts_id_idx", "yts_id"),
        Index("title_idx", "title"),
    )

# ---------------------------------------------------------
# MOVIE DIRECTORS TABLE (NEW)
# ---------------------------------------------------------
class MovieDirector(Base):
    __tablename__ = "movie_directors"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    movie_id = Column(UUID(as_uuid=True), ForeignKey("movies.id", ondelete="CASCADE"), nullable=False)
    name = Column(String, nullable=False)
    image_url = Column(String)

    movie = relationship("Movie", back_populates="directors")

    __table_args__ = (
        Index("director_movie_id_idx", "movie_id"),
    )

# ---------------------------------------------------------
# MOVIE CAST TABLE (NEW)
# ---------------------------------------------------------
class MovieCast(Base):
    __tablename__ = "movie_cast"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    movie_id = Column(UUID(as_uuid=True), ForeignKey("movies.id", ondelete="CASCADE"), nullable=False)
    name = Column(String, nullable=False)
    character_name = Column(String)
    image_url = Column(String)
    imdb_code = Column(String)

    movie = relationship("Movie", back_populates="cast")

    __table_args__ = (
        Index("cast_movie_id_idx", "movie_id"),
    )

# ---------------------------------------------------------
# TORRENTS TABLE
# ---------------------------------------------------------
class Torrent(Base):
    __tablename__ = "torrents"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    movie_id = Column(UUID(as_uuid=True), ForeignKey("movies.id", ondelete="CASCADE"), nullable=False)

    quality = Column(String, nullable=False)
    type = Column(String, nullable=False)
    is_repack = Column(Boolean, default=False)
    is_proper = Column(Boolean, default=False)

    file_size_bytes = Column(BigInteger)
    file_size_human = Column(String)

    torrent_hash = Column(String, unique=True, nullable=False)
    torrent_file_url = Column(String, nullable=False)
    minio_torrent_path = Column(String)

    seeds = Column(Integer, default=0)
    peers = Column(Integer, default=0)

    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False)

    movie = relationship("Movie", back_populates="torrents")
    downloads = relationship("Download", back_populates="torrent")

    __table_args__ = (
        UniqueConstraint("movie_id", "quality", "type", name="movieQualityUnique"),
        Index("torrent_hash_idx", "torrent_hash"),
    )

# ---------------------------------------------------------
# DOWNLOADS TABLE
# ---------------------------------------------------------
class Download(Base):
    __tablename__ = "downloads"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    torrent_id = Column(UUID(as_uuid=True), ForeignKey("torrents.id", ondelete="RESTRICT"), nullable=False)
    movie_id = Column(UUID(as_uuid=True), ForeignKey("movies.id", ondelete="CASCADE"), nullable=False)
    session_id = Column(UUID(as_uuid=True), unique=True, nullable=False)

    source_type = Column(Enum(SourceType, name="source_type"), nullable=False)
    source_uri = Column(String, nullable=False)

    state = Column(Enum(DownloadState, name="download_state"), default=DownloadState.IDLE, nullable=False)
    progress = Column(Float, default=0.0, nullable=False)
    downloaded_bytes = Column(BigInteger, default=0)
    total_bytes = Column(BigInteger, default=0)
    error_message = Column(String)

    storage_provider = Column(Enum(StorageProvider, name="storage_provider"), default=StorageProvider.NONE)
    local_path = Column(String)

    video_codec = Column(String)
    audio_codec = Column(String)
    audio_channels = Column(String)
    resolution_width = Column(Integer)
    resolution_height = Column(Integer)
    duration_seconds = Column(Integer)

    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False)

    movie = relationship("Movie", back_populates="downloads")
    torrent = relationship("Torrent", back_populates="downloads")
    subtitles = relationship("Subtitle", back_populates="download", cascade="all, delete-orphan")

# ---------------------------------------------------------
# SUBTITLES TABLE
# ---------------------------------------------------------
class Subtitle(Base):
    __tablename__ = "subtitles"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    download_id = Column(UUID(as_uuid=True), ForeignKey("downloads.id", ondelete="CASCADE"), nullable=False)

    language_code = Column(String, nullable=False)
    language_name = Column(String)
    is_forced = Column(Boolean, default=False)
    is_sdh = Column(Boolean, default=False)
    original_filename = Column(String)
    minio_path = Column(String, nullable=False)

    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)

    download = relationship("Download", back_populates="subtitles")

# ---------------------------------------------------------
# USER PLAYBACKS TABLE
# ---------------------------------------------------------
class UserPlayback(Base):
    __tablename__ = "user_playbacks"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id = Column(String, nullable=False)
    download_id = Column(UUID(as_uuid=True), ForeignKey("downloads.id", ondelete="CASCADE"), nullable=False)
    movie_id = Column(UUID(as_uuid=True), ForeignKey("movies.id", ondelete="CASCADE"), nullable=False)
    
    last_watched_seconds = Column(Float, default=0)
    
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False)

    movie = relationship("Movie", back_populates="playbacks")

    __table_args__ = (
        UniqueConstraint("user_id", "download_id", name="user_download_unq"),
    )

class WatchHistory(Base):
    __tablename__ = "watch_history"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id = Column(String, nullable=False)
    movie_id = Column(UUID(as_uuid=True), ForeignKey("movies.id", ondelete="CASCADE"), nullable=False)

    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False)

    movie = relationship("Movie", back_populates="watch_history_entries")

    __table_args__ = (
        Index("wh_user_movie_idx", "user_id", "movie_id"),
    )

class UserFavorite(Base):
    __tablename__ = "user_favorites"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id = Column(String, nullable=False)
    movie_id = Column(UUID(as_uuid=True), ForeignKey("movies.id", ondelete="CASCADE"), nullable=False)

    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)

    movie = relationship("Movie", back_populates="favorites")

    __table_args__ = (
        UniqueConstraint("user_id", "movie_id", name="fav_user_movie_unq"),
        Index("fav_user_id_idx", "user_id"),
    )

class MovieComment(Base):
    __tablename__ = "movie_comments"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    movie_id = Column(UUID(as_uuid=True), ForeignKey("movies.id", ondelete="CASCADE"), nullable=False)
    user_id = Column(String, nullable=False)
    username = Column(String, nullable=False)
    content = Column(String, nullable=False)

    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False)

    movie = relationship("Movie", back_populates="comments")

    __table_args__ = (
        Index("comment_movie_id_idx", "movie_id"),
        Index("comment_created_at_idx", "created_at"),
    )

File: services/hypertube/drizzle/schema.ts
import { relations } from "drizzle-orm";
import {
    pgTable,
    text,
    integer,
    doublePrecision,
    timestamp,
    uuid,
    pgEnum,
    unique,
    bigint,
    boolean,
    index,
} from "drizzle-orm/pg-core";

// ---------------------------------------------------------
// ENUMS
// ---------------------------------------------------------
export const downloadStateEnum = pgEnum("download_state", [
    "IDLE",
    "FETCHING_METADATA",
    "DOWNLOADING",
    "PROCESSING",
    "COMPLETED",
    "ERROR",
]);

export const storageProviderEnum = pgEnum("storage_provider", [
    "NONE",
    "LOCAL_DISK",
    "MINIO",
]);

export const sourceTypeEnum = pgEnum("source_type", ["MAGNET", "TORRENT_FILE"]);

// ---------------------------------------------------------
// MOVIES TABLE
// ---------------------------------------------------------
export const movies = pgTable(
    "movies",
    {
        id: uuid("id").defaultRandom().primaryKey(),
        ytsId: integer("yts_id").unique().notNull(),
        title: text("title").notNull(),
        titleLong: text("title_long"),
        releaseYear: integer("release_year").notNull(),
        rating: doublePrecision("rating"),
        runtimeMinutes: integer("runtime_minutes"),
        description: text("description"),
        genres: text("genres").array(),
        language: text("language"),
        mpaRating: text("mpa_rating"),
        ytsPosterUrl: text("yts_poster_url").notNull(),
        minioPosterPath: text("minio_poster_path"),
        ytsBackgroundImageUrl: text("yts_background_image_url"),
        minioBackgroundImagePath: text("minio_background_image_path"),
        imdbCode: text("imdb_code"),
        dateUploaded: timestamp("date_uploaded", { withTimezone: true }),

        // --- NEW FIELDS ---
        isNewRelease: boolean("is_new_release").default(false),
        isPopular: boolean("is_popular").default(false),

        lastCheckedAt: timestamp("last_checked_at", { withTimezone: true })
            .defaultNow()
            .notNull(),

        createdAt: timestamp("created_at", { withTimezone: true })
            .defaultNow()
            .notNull(),
        updatedAt: timestamp("updated_at", { withTimezone: true })
            .defaultNow()
            .$onUpdate(() => new Date())
            .notNull(),
    },
    (table) => ({
        ytsIdIdx: index("yts_id_idx").on(table.ytsId),
        titleIdx: index("title_idx").on(table.title),
    })
);

// ---------------------------------------------------------
// MOVIE DIRECTORS TABLE (NEW)
// ---------------------------------------------------------
export const movieDirectors = pgTable(
    "movie_directors",
    {
        id: uuid("id").defaultRandom().primaryKey(),
        movieId: uuid("movie_id")
            .references(() => movies.id, { onDelete: "cascade" })
            .notNull(),
        name: text("name").notNull(),
        imageUrl: text("image_url"),
    },
    (table) => ({
        movieIdIdx: index("director_movie_id_idx").on(table.movieId),
    })
);

// ---------------------------------------------------------
// MOVIE CAST TABLE (NEW)
// ---------------------------------------------------------
export const movieCast = pgTable(
    "movie_cast",
    {
        id: uuid("id").defaultRandom().primaryKey(),
        movieId: uuid("movie_id")
            .references(() => movies.id, { onDelete: "cascade" })
            .notNull(),
        name: text("name").notNull(),
        characterName: text("character_name"),
        imageUrl: text("image_url"),
        imdbCode: text("imdb_code"),
    },
    (table) => ({
        movieIdIdx: index("cast_movie_id_idx").on(table.movieId),
    })
);

// ---------------------------------------------------------
// TORRENTS TABLE
// ---------------------------------------------------------
export const torrents = pgTable(
    "torrents",
    {
        id: uuid("id").defaultRandom().primaryKey(),
        movieId: uuid("movie_id")
            .references(() => movies.id, { onDelete: "cascade" })
            .notNull(),
        quality: text("quality").notNull(),
        type: text("type").notNull(),
        isRepack: boolean("is_repack").default(false),
        isProper: boolean("is_proper").default(false),
        fileSizeBytes: bigint("file_size_bytes", { mode: "number" }),
        fileSizeHuman: text("file_size_human"),
        torrentHash: text("torrent_hash").unique().notNull(),
        torrentFileUrl: text("torrent_file_url").notNull(),
        minioTorrentPath: text("minio_torrent_path"),
        seeds: integer("seeds").default(0),
        peers: integer("peers").default(0),
        createdAt: timestamp("created_at", { withTimezone: true })
            .defaultNow()
            .notNull(),
        updatedAt: timestamp("updated_at", { withTimezone: true })
            .defaultNow()
            .$onUpdate(() => new Date())
            .notNull(),
    },
    (table) => ({
        movieQualityUnique: unique("movieQualityUnique").on(
            table.movieId,
            table.quality,
            table.type
        ),
        hashIdx: index("torrent_hash_idx").on(table.torrentHash),
    })
);

// ---------------------------------------------------------
// DOWNLOADS TABLE
// ---------------------------------------------------------
export const downloads = pgTable("downloads", {
    id: uuid("id").defaultRandom().primaryKey(),
    torrentId: uuid("torrent_id")
        .references(() => torrents.id, { onDelete: "restrict" })
        .notNull(),
    movieId: uuid("movie_id")
        .references(() => movies.id, { onDelete: "cascade" })
        .notNull(),
    sessionId: uuid("session_id").unique().notNull(),
    sourceType: sourceTypeEnum("source_type").notNull(),
    sourceUri: text("source_uri").notNull(),
    state: downloadStateEnum("state").default("IDLE").notNull(),
    progress: doublePrecision("progress").default(0.0).notNull(),
    downloadedBytes: bigint("downloaded_bytes", { mode: "number" }).default(0),
    totalBytes: bigint("total_bytes", { mode: "number" }).default(0),
    errorMessage: text("error_message"),
    storageProvider: storageProviderEnum("storage_provider").default("NONE"),
    localPath: text("local_path"),
    videoCodec: text("video_codec"),
    audioCodec: text("audio_codec"),
    audioChannels: text("audio_channels"),
    resolutionWidth: integer("resolution_width"),
    resolutionHeight: integer("resolution_height"),
    durationSeconds: integer("duration_seconds"),
    createdAt: timestamp("created_at", { withTimezone: true })
        .defaultNow()
        .notNull(),
    updatedAt: timestamp("updated_at", { withTimezone: true })
        .defaultNow()
        .$onUpdate(() => new Date())
        .notNull(),
});

// ---------------------------------------------------------
// SUBTITLES TABLE
// ---------------------------------------------------------
export const subtitles = pgTable("subtitles", {
    id: uuid("id").defaultRandom().primaryKey(),
    downloadId: uuid("download_id")
        .references(() => downloads.id, { onDelete: "cascade" })
        .notNull(),
    languageCode: text("language_code").notNull(),
    languageName: text("language_name"),
    isForced: boolean("is_forced").default(false),
    isSdh: boolean("is_sdh").default(false),
    originalFilename: text("original_filename"),
    minioPath: text("minio_path").notNull(),
    createdAt: timestamp("created_at", { withTimezone: true })
        .defaultNow()
        .notNull(),
});

// ---------------------------------------------------------
// USER PLAYBACKS TABLE
// ---------------------------------------------------------
export const userPlaybacks = pgTable(
    "user_playbacks",
    {
        id: uuid("id").defaultRandom().primaryKey(),
        userId: text("user_id").notNull(),
        downloadId: uuid("download_id")
            .references(() => downloads.id, { onDelete: "cascade" })
            .notNull(),
        movieId: uuid("movie_id")
            .references(() => movies.id, { onDelete: "cascade" })
            .notNull(),
        lastWatchedSeconds: doublePrecision("last_watched_seconds").default(0),
        updatedAt: timestamp("updated_at", { withTimezone: true })
            .defaultNow()
            .$onUpdate(() => new Date())
            .notNull(),
    },
    (table) => ({
        userDownloadUnq: unique("user_download_unq").on(
            table.userId,
            table.downloadId
        ),
    })
);

export const watchHistory = pgTable(
    "watch_history",
    {
        id: uuid("id").defaultRandom().primaryKey(),
        userId: text("user_id").notNull(),
        movieId: uuid("movie_id")
            .references(() => movies.id, { onDelete: "cascade" })
            .notNull(),
        createdAt: timestamp("created_at", { withTimezone: true })
            .defaultNow()
            .notNull(),
        updatedAt: timestamp("updated_at", { withTimezone: true })
            .defaultNow()
            .$onUpdate(() => new Date())
            .notNull(),
    },
    (table) => ({
        userMovieIdx: index("wh_user_movie_idx").on(
            table.userId,
            table.movieId
        ),
    })
);

export const userFavorites = pgTable(
    "user_favorites",
    {
        id: uuid("id").defaultRandom().primaryKey(),
        userId: text("user_id").notNull(),
        movieId: uuid("movie_id")
            .references(() => movies.id, { onDelete: "cascade" })
            .notNull(),
        createdAt: timestamp("created_at", { withTimezone: true })
            .defaultNow()
            .notNull(),
    },
    (table) => ({
        userMovieUnq: unique("fav_user_movie_unq").on(
            table.userId,
            table.movieId
        ),
        userIdIdx: index("fav_user_id_idx").on(table.userId),
    })
);

export const movieComments = pgTable(
    "movie_comments",
    {
        id: uuid("id").defaultRandom().primaryKey(),
        movieId: uuid("movie_id")
            .references(() => movies.id, { onDelete: "cascade" })
            .notNull(),
        userId: text("user_id").notNull(),
        username: text("username").notNull(),
        content: text("content").notNull(),
        createdAt: timestamp("created_at", { withTimezone: true })
            .defaultNow()
            .notNull(),
        updatedAt: timestamp("updated_at", { withTimezone: true })
            .defaultNow()
            .$onUpdate(() => new Date())
            .notNull(),
    },
    (table) => ({
        movieIdIdx: index("comment_movie_id_idx").on(table.movieId),
        createdAtIdx: index("comment_created_at_idx").on(table.createdAt),
    })
);

// ---------------------------------------------------------
// RELATIONS
// ---------------------------------------------------------

export const moviesRelations = relations(movies, ({ many }) => ({
    torrents: many(torrents),
    downloads: many(downloads),
    playbacks: many(userPlaybacks),
    directors: many(movieDirectors),
    cast: many(movieCast),
    watchHistory: many(watchHistory),
    favorites: many(userFavorites),
    comments: many(movieComments),
}));

// NEW RELATIONS BLOCK
export const movieDirectorsRelations = relations(movieDirectors, ({ one }) => ({
    movie: one(movies, {
        fields: [movieDirectors.movieId],
        references: [movies.id],
    }),
}));

// NEW RELATIONS BLOCK
export const movieCastRelations = relations(movieCast, ({ one }) => ({
    movie: one(movies, {
        fields: [movieCast.movieId],
        references: [movies.id],
    }),
}));

export const torrentsRelations = relations(torrents, ({ one, many }) => ({
    movie: one(movies, {
        fields: [torrents.movieId],
        references: [movies.id],
    }),
    downloads: many(downloads),
}));

export const downloadsRelations = relations(downloads, ({ one, many }) => ({
    movie: one(movies, {
        fields: [downloads.movieId],
        references: [movies.id],
    }),
    torrent: one(torrents, {
        fields: [downloads.torrentId],
        references: [torrents.id],
    }),
    subtitles: many(subtitles),
}));

export const subtitlesRelations = relations(subtitles, ({ one }) => ({
    download: one(downloads, {
        fields: [subtitles.downloadId],
        references: [downloads.id],
    }),
}));

export const userPlaybacksRelations = relations(userPlaybacks, ({ one }) => ({
    movie: one(movies, {
        fields: [userPlaybacks.movieId],
        references: [movies.id],
    }),
}));

export const watchHistoryRelations = relations(watchHistory, ({ one }) => ({
    movie: one(movies, {
        fields: [watchHistory.movieId],
        references: [movies.id],
    }),
}));

export const userFavoritesRelations = relations(userFavorites, ({ one }) => ({
    movie: one(movies, {
        fields: [userFavorites.movieId],
        references: [movies.id],
    }),
}));

export const movieCommentsRelations = relations(movieComments, ({ one }) => ({
    movie: one(movies, {
        fields: [movieComments.movieId],
        references: [movies.id],
    }),
}));
